<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a domino object and prepare it for network construction — create_domino • dominoSignal</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Fira_Code-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a domino object and prepare it for network construction — create_domino"><meta name="description" content="This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for build_domino(), which will calculate a signaling
network."><meta property="og:description" content="This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for build_domino(), which will calculate a signaling
network."><meta property="og:image" content="https://FertigLab.github.io/dominoSignal/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dominoSignal</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/dominoSignal.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/cellphonedb_vignette.html">Using the CellPhoneDB Database</a></li>
    <li><a class="dropdown-item" href="../articles/domino_object_vignette.html">Interacting with domino Objects</a></li>
    <li><a class="dropdown-item" href="../articles/plotting_vignette.html">Plotting Functions and Options</a></li>
    <li><a class="dropdown-item" href="../articles/tf_scenic_vignette.html">SCENIC for TF Activation Scoring</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-news" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">News</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-news"><li><h6 class="dropdown-header" data-toc-skip>Releases</h6></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.1-alpha">dominoSignal v0.2.1</a></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.2-alpha">dominoSignal v0.2.2</a></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.99.2-alpha">dominoSignal v0.99.2</a></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v1.0.0-Bioconductor">dominoSignal v1.0.0 for Bioconductor</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../news/index.html">Changelog</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/FertigLab/dominoSignal/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Create a domino object and prepare it for network construction</h1>
      <small class="dont-index">Source: <a href="https://github.com/FertigLab/dominoSignal/blob/master/R/import_fxns.R" class="external-link"><code>R/import_fxns.R</code></a></small>
      <div class="d-none name"><code>create_domino.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for <code><a href="build_domino.html">build_domino()</a></code>, which will calculate a signaling
network.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_domino</span><span class="op">(</span></span>
<span>  <span class="va">rl_map</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  z_scores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clusters <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_clusters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tf_targets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_complexes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rec_min_thresh <span class="op">=</span> <span class="fl">0.025</span>,</span>
<span>  remove_rec_dropout <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tf_selection_method <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  tf_variance_quantile <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-rl-map">rl_map<a class="anchor" aria-label="anchor" href="#arg-rl-map"></a></dt>
<dd><p>Data frame where each row describes a receptor-ligand interaction with required columns gene_A &amp; gene_B including the gene names for the receptor and ligand and type_A &amp; type_B annotating if genes A and B are a ligand (L) or receptor (R)</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Either a path to a csv containing cell level features of interest (ie. the auc matrix from pySCENIC) or named matrix with cells as columns and features as rows.</p></dd>


<dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>Counts matrix for the data. This is only used to threshold receptors on dropout.</p></dd>


<dt id="arg-z-scores">z_scores<a class="anchor" aria-label="anchor" href="#arg-z-scores"></a></dt>
<dd><p>Matrix containing z-scored expression data for all cells with cells as columns and features as rows.</p></dd>


<dt id="arg-clusters">clusters<a class="anchor" aria-label="anchor" href="#arg-clusters"></a></dt>
<dd><p>Named factor containing cell cluster with names as cells.</p></dd>


<dt id="arg-use-clusters">use_clusters<a class="anchor" aria-label="anchor" href="#arg-use-clusters"></a></dt>
<dd><p>Boolean indicating whether to use clusters.</p></dd>


<dt id="arg-tf-targets">tf_targets<a class="anchor" aria-label="anchor" href="#arg-tf-targets"></a></dt>
<dd><p>Optional. A list where names are transcription factors and the stored values are character vectors of genes in the transcription factor's regulon.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean indicating whether or not to print progress during computation.</p></dd>


<dt id="arg-use-complexes">use_complexes<a class="anchor" aria-label="anchor" href="#arg-use-complexes"></a></dt>
<dd><p>Boolean indicating whether you wish to use receptor/ligand complexes in the receptor ligand signaling database. If FALSE, receptor/ligand pairs where either functions as a protein complex will not be considered when constructing the signaling network.</p></dd>


<dt id="arg-rec-min-thresh">rec_min_thresh<a class="anchor" aria-label="anchor" href="#arg-rec-min-thresh"></a></dt>
<dd><p>Minimum expression level of receptors by cell. Default is 0.025 or 2.5 percent of all cells in the data set. This is important when calculating correlation to connect receptors to transcription activation. If this threshold is too low then correlation calculations will proceed with very few cells with non-zero expression.</p></dd>


<dt id="arg-remove-rec-dropout">remove_rec_dropout<a class="anchor" aria-label="anchor" href="#arg-remove-rec-dropout"></a></dt>
<dd><p>Whether to remove receptors with 0 expression counts when calculating correlations. This can reduce false positive correlation calculations when receptors have high dropout rates.</p></dd>


<dt id="arg-tf-selection-method">tf_selection_method<a class="anchor" aria-label="anchor" href="#arg-tf-selection-method"></a></dt>
<dd><p>Selection of which method to target transcription factors. If 'clusters' then differential expression for clusters will be calculated. If 'variable' then the most variable transcription factors will be selected. If 'all' then all transcription factors in the feature matrix will be used. Default is 'clusters'. Note that if you wish to use clusters for intercellular signaling downstream to MUST choose clusters.</p></dd>


<dt id="arg-tf-variance-quantile">tf_variance_quantile<a class="anchor" aria-label="anchor" href="#arg-tf-variance-quantile"></a></dt>
<dd><p>What proportion of variable features to take if using variance to threshold features. Default is 0.5. Higher numbers will keep more features. Ignored if tf_selection_method is not 'variable'</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A domino object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">create_rl_map_cellphonedb</span>, echo <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">create_regulon_list_scenic</span>, echo <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">SCENIC</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pbmc_dom_tiny</span> <span class="op">&lt;-</span> <span class="fu">create_domino</span><span class="op">(</span></span></span>
<span class="r-in"><span> rl_map <span class="op">=</span> <span class="va">rl_map_tiny</span>, features <span class="op">=</span> <span class="va">SCENIC</span><span class="op">$</span><span class="va">auc_tiny</span>,</span></span>
<span class="r-in"><span> counts <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_count_tiny</span>, z_scores <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_zscore_tiny</span>,</span></span>
<span class="r-in"><span> clusters <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">clusters_tiny</span>, tf_targets <span class="op">=</span> <span class="va">regulon_list_tiny</span>,</span></span>
<span class="r-in"><span> use_clusters <span class="op">=</span> <span class="cn">TRUE</span>, use_complexes <span class="op">=</span> <span class="cn">TRUE</span>, remove_rec_dropout <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span> verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pbmc_dom_tiny_no_clusters</span> <span class="op">&lt;-</span> <span class="fu">create_domino</span><span class="op">(</span></span></span>
<span class="r-in"><span> rl_map <span class="op">=</span> <span class="va">rl_map_tiny</span>, features <span class="op">=</span> <span class="va">SCENIC</span><span class="op">$</span><span class="va">auc_tiny</span>,</span></span>
<span class="r-in"><span> counts <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_count_tiny</span>, z_scores <span class="op">=</span><span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_zscore_tiny</span>,</span></span>
<span class="r-in"><span> clusters <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">clusters_tiny</span>, tf_targets <span class="op">=</span> <span class="va">regulon_list_tiny</span>,</span></span>
<span class="r-in"><span> use_clusters <span class="op">=</span> <span class="cn">FALSE</span>, use_complexes <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span> rec_min_thresh <span class="op">=</span> <span class="fl">0.1</span>, remove_rec_dropout <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span> tf_selection_method <span class="op">=</span> <span class="st">"all"</span>,</span></span>
<span class="r-in"><span> verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>dominoSignal is developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan, Dmitrijs Lvovs.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

