<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a receptor - ligand map from a CellPhoneDB signaling database — create_rl_map_cellphonedb • dominoSignal</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Fira_Code-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a receptor - ligand map from a CellPhoneDB signaling database — create_rl_map_cellphonedb"><meta name="description" content="Generates a data frame of ligand-receptor interactions from a CellPhoneDB database annotating the genes encoding the interacting ligands and receptors to be queried in transcriptomic data."><meta property="og:description" content="Generates a data frame of ligand-receptor interactions from a CellPhoneDB database annotating the genes encoding the interacting ligands and receptors to be queried in transcriptomic data."><meta property="og:image" content="https://FertigLab.github.io/dominoSignal/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dominoSignal</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/dominoSignal.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/cellphonedb_vignette.html">Using the CellPhoneDB Database</a></li>
    <li><a class="dropdown-item" href="../articles/domino_object_vignette.html">Interacting with domino Objects</a></li>
    <li><a class="dropdown-item" href="../articles/plotting_vignette.html">Plotting Functions and Options</a></li>
    <li><a class="dropdown-item" href="../articles/tf_scenic_vignette.html">SCENIC for TF Activation Scoring</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-news" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">News</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-news"><li><h6 class="dropdown-header" data-toc-skip>Releases</h6></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.1-alpha">dominoSignal v0.2.1</a></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.2.2-alpha">dominoSignal v0.2.2</a></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v0.99.2-alpha">dominoSignal v0.99.2</a></li>
    <li><a class="external-link dropdown-item" href="https://github.com/FertigLab/dominoSignal/releases/tag/v1.0.0-Bioconductor">dominoSignal v1.0.0 for Bioconductor</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../news/index.html">Changelog</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/FertigLab/dominoSignal/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Create a receptor - ligand map from a CellPhoneDB signaling database</h1>
      <small class="dont-index">Source: <a href="https://github.com/FertigLab/dominoSignal/blob/master/R/import_fxns.R" class="external-link"><code>R/import_fxns.R</code></a></small>
      <div class="d-none name"><code>create_rl_map_cellphonedb.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generates a data frame of ligand-receptor interactions from a CellPhoneDB database annotating the genes encoding the interacting ligands and receptors to be queried in transcriptomic data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_rl_map_cellphonedb</span><span class="op">(</span></span>
<span>  <span class="va">genes</span>,</span>
<span>  <span class="va">proteins</span>,</span>
<span>  <span class="va">interactions</span>,</span>
<span>  complexes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  database_name <span class="op">=</span> <span class="st">"CellPhoneDB"</span>,</span>
<span>  gene_conv <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_conv_host <span class="op">=</span> <span class="st">"https://www.ensembl.org"</span>,</span>
<span>  alternate_convert <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  alternate_convert_table <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>data frame or file path to table of gene names in uniprot, hgnc_symbol, or ensembl format in CellPhoneDB database format</p></dd>


<dt id="arg-proteins">proteins<a class="anchor" aria-label="anchor" href="#arg-proteins"></a></dt>
<dd><p>data frame or file path to table of protein features in CellPhoneDB format</p></dd>


<dt id="arg-interactions">interactions<a class="anchor" aria-label="anchor" href="#arg-interactions"></a></dt>
<dd><p>data frame or file path to table of protein-protein interactions in CellPhoneDB format</p></dd>


<dt id="arg-complexes">complexes<a class="anchor" aria-label="anchor" href="#arg-complexes"></a></dt>
<dd><p>optional: data frame or file path to table of protein complexes in CellPhoneDB format</p></dd>


<dt id="arg-database-name">database_name<a class="anchor" aria-label="anchor" href="#arg-database-name"></a></dt>
<dd><p>name of the database being used, stored in output</p></dd>


<dt id="arg-gene-conv">gene_conv<a class="anchor" aria-label="anchor" href="#arg-gene-conv"></a></dt>
<dd><p>a tuple of (from, to) or (source, target) if gene conversion to orthologs is desired; options are ENSMUSG, ENSG, MGI, or HGNC</p></dd>


<dt id="arg-gene-conv-host">gene_conv_host<a class="anchor" aria-label="anchor" href="#arg-gene-conv-host"></a></dt>
<dd><p>host for conversion; default ensembl, could also use mirrors if desired</p></dd>


<dt id="arg-alternate-convert">alternate_convert<a class="anchor" aria-label="anchor" href="#arg-alternate-convert"></a></dt>
<dd><p>boolean if you would like to use a non-ensembl method of conversion (must supply table; not recommended, use only if ensembl is down)</p></dd>


<dt id="arg-alternate-convert-table">alternate_convert_table<a class="anchor" aria-label="anchor" href="#arg-alternate-convert-table"></a></dt>
<dd><p>supplied table for non-ensembl method of conversion</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Data frame where each row describes a possible receptor-ligand interaction</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CellPhoneDB</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rl_map_tiny</span> <span class="op">&lt;-</span> <span class="fu">create_rl_map_cellphonedb</span><span class="op">(</span>genes <span class="op">=</span> <span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">genes_tiny</span>,</span></span>
<span class="r-in"><span> proteins <span class="op">=</span> <span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">proteins_tiny</span>,</span></span>
<span class="r-in"><span> interactions <span class="op">=</span> <span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">interactions_tiny</span>,</span></span>
<span class="r-in"><span> complexes <span class="op">=</span><span class="va">CellPhoneDB</span><span class="op">$</span><span class="va">complexes_tiny</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>dominoSignal is developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan, Dmitrijs Lvovs.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

