<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create a domino object and prepare it for network construction — create_domino • dominoSignal</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create a domino object and prepare it for network construction — create_domino"><meta property="og:description" content="This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for build_domino(), which will calculate a signaling
network."><meta property="og:image" content="/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dominoSignal</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/dominoSignal.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cellphonedb_vignette.html">Using the CellPhoneDB Database</a>
    </li>
    <li>
      <a href="../articles/domino_object_vignette.html">Interacting with domino Objects</a>
    </li>
    <li>
      <a href="../articles/plotting_vignette.html">Plotting Functions and Options</a>
    </li>
    <li>
      <a href="../articles/tf_scenic_vignette.html">SCENIC for TF Activation Scoring</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/FertigLab/dominoSignal/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create a domino object and prepare it for network construction</h1>
    <small class="dont-index">Source: <a href="https://github.com/FertigLab/dominoSignal/blob/master/R/import_fxns.R" class="external-link"><code>R/import_fxns.R</code></a></small>
    <div class="hidden name"><code>create_domino.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function reads in a receptor ligand signaling database, cell level
features of some kind (ie. output from pySCENIC), z-scored single cell data,
and cluster id for single cell data, calculates a correlation matrix between
receptors and other features (this is transcription factor module scores if
using pySCENIC), and finds features enriched by cluster. It will return a
domino object prepared for <code><a href="build_domino.html">build_domino()</a></code>, which will calculate a signaling
network.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_domino</span><span class="op">(</span></span>
<span>  <span class="va">rl_map</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  z_scores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clusters <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_clusters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tf_targets <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  use_complexes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rec_min_thresh <span class="op">=</span> <span class="fl">0.025</span>,</span>
<span>  remove_rec_dropout <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tf_selection_method <span class="op">=</span> <span class="st">"clusters"</span>,</span>
<span>  tf_variance_quantile <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-rl-map">rl_map<a class="anchor" aria-label="anchor" href="#arg-rl-map"></a></dt>
<dd><p>Data frame where each row describes a receptor-ligand interaction with required columns gene_A &amp; gene_B including the gene names for the receptor and ligand and type_A &amp; type_B annotating if genes A and B are a ligand (L) or receptor (R)</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Either a path to a csv containing cell level features of interest (ie. the auc matrix from pySCENIC) or named matrix with cells as columns and features as rows.</p></dd>


<dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>Counts matrix for the data. This is only used to threshold receptors on dropout.</p></dd>


<dt id="arg-z-scores">z_scores<a class="anchor" aria-label="anchor" href="#arg-z-scores"></a></dt>
<dd><p>Matrix containing z-scored expression data for all cells with cells as columns and features as rows.</p></dd>


<dt id="arg-clusters">clusters<a class="anchor" aria-label="anchor" href="#arg-clusters"></a></dt>
<dd><p>Named factor containing cell cluster with names as cells.</p></dd>


<dt id="arg-use-clusters">use_clusters<a class="anchor" aria-label="anchor" href="#arg-use-clusters"></a></dt>
<dd><p>Boolean indicating whether to use clusters.</p></dd>


<dt id="arg-tf-targets">tf_targets<a class="anchor" aria-label="anchor" href="#arg-tf-targets"></a></dt>
<dd><p>Optional. A list where names are transcription factors and the stored values are character vectors of genes in the transcription factor's regulon.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean indicating whether or not to print progress during computation.</p></dd>


<dt id="arg-use-complexes">use_complexes<a class="anchor" aria-label="anchor" href="#arg-use-complexes"></a></dt>
<dd><p>Boolean indicating whether you wish to use receptor/ligand complexes in the receptor ligand signaling database. If FALSE, receptor/ligand pairs where either functions as a protein complex will not be considered when constructing the signaling network.</p></dd>


<dt id="arg-rec-min-thresh">rec_min_thresh<a class="anchor" aria-label="anchor" href="#arg-rec-min-thresh"></a></dt>
<dd><p>Minimum expression level of receptors by cell. Default is 0.025 or 2.5 percent of all cells in the data set. This is important when calculating correlation to connect receptors to transcription activation. If this threshold is too low then correlation calculations will proceed with very few cells with non-zero expression.</p></dd>


<dt id="arg-remove-rec-dropout">remove_rec_dropout<a class="anchor" aria-label="anchor" href="#arg-remove-rec-dropout"></a></dt>
<dd><p>Whether to remove receptors with 0 expression counts when calculating correlations. This can reduce false positive correlation calculations when receptors have high dropout rates.</p></dd>


<dt id="arg-tf-selection-method">tf_selection_method<a class="anchor" aria-label="anchor" href="#arg-tf-selection-method"></a></dt>
<dd><p>Selection of which method to target transcription factors. If 'clusters' then differential expression for clusters will be calculated. If 'variable' then the most variable transcription factors will be selected. If 'all' then all transcription factors in the feature matrix will be used. Default is 'clusters'. Note that if you wish to use clusters for intercellular signaling downstream to MUST choose clusters.</p></dd>


<dt id="arg-tf-variance-quantile">tf_variance_quantile<a class="anchor" aria-label="anchor" href="#arg-tf-variance-quantile"></a></dt>
<dd><p>What proportion of variable features to take if using variance to threshold features. Default is 0.5. Higher numbers will keep more features. Ignored if tf_selection_method is not 'variable'</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A domino object</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">create_rl_map_cellphonedb</span>, echo <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">create_regulon_list_scenic</span>, echo <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">SCENIC</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pbmc_dom_tiny</span> <span class="op">&lt;-</span> <span class="fu">create_domino</span><span class="op">(</span></span></span>
<span class="r-in"><span> rl_map <span class="op">=</span> <span class="va">rl_map_tiny</span>, features <span class="op">=</span> <span class="va">SCENIC</span><span class="op">$</span><span class="va">auc_tiny</span>,</span></span>
<span class="r-in"><span> counts <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_count_tiny</span>, z_scores <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_zscore_tiny</span>,</span></span>
<span class="r-in"><span> clusters <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">clusters_tiny</span>, tf_targets <span class="op">=</span> <span class="va">regulon_list_tiny</span>,</span></span>
<span class="r-in"><span> use_clusters <span class="op">=</span> <span class="cn">TRUE</span>, use_complexes <span class="op">=</span> <span class="cn">TRUE</span>, remove_rec_dropout <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span> verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pbmc_dom_tiny_no_clusters</span> <span class="op">&lt;-</span> <span class="fu">create_domino</span><span class="op">(</span></span></span>
<span class="r-in"><span> rl_map <span class="op">=</span> <span class="va">rl_map_tiny</span>, features <span class="op">=</span> <span class="va">SCENIC</span><span class="op">$</span><span class="va">auc_tiny</span>,</span></span>
<span class="r-in"><span> counts <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_count_tiny</span>, z_scores <span class="op">=</span><span class="va">PBMC</span><span class="op">$</span><span class="va">RNA_zscore_tiny</span>,</span></span>
<span class="r-in"><span> clusters <span class="op">=</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">clusters_tiny</span>, tf_targets <span class="op">=</span> <span class="va">regulon_list_tiny</span>,</span></span>
<span class="r-in"><span> use_clusters <span class="op">=</span> <span class="cn">FALSE</span>, use_complexes <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span> rec_min_thresh <span class="op">=</span> <span class="fl">0.1</span>, remove_rec_dropout <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span> tf_selection_method <span class="op">=</span> <span class="st">"all"</span>,</span></span>
<span class="r-in"><span> verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Christopher Cherry, Jacob T Mitchell, Sushma Nagaraj, Kavita Krishnan, Dmitrijs Lvovs.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

